src/libxively/mqtt/codec/xi_mqtt_codec_layer.c

```diff
@@ -277,10 +277,16 @@ xi_state_t xi_mqtt_codec_layer_pull( void* context, void* data, xi_state_t in_ou
         layer_data->local_state =
             xi_mqtt_parser_execute( &layer_data->parser, layer_data->msg, data_desc );
 
+        xi_debug_format( "local_state = %d", ( int )layer_data->local_state );
+
+        if ( layer_data->local_state == XI_STATE_WANT_READ )
+        {
+            xi_free_desc( &data_desc );
```

Comment:
ok, tests MUST come here to cover these. The story should include these integration tests as well!